patches/server/0962-Add-player-death-screen-message-API.patch

```diff
@@ -0,0 +1,53 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: dawon <dawon@dawon.eu>
+Date: Fri, 24 Feb 2023 15:21:21 +0100
+Subject: [PATCH] Add player death screen message API
+
+
+diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
+index 9160725f031c9698a00229c3dfa19d39b4a826c1..f91add6c8d21fd84a4b9f4c40c4f8c62cc5361d0 100644
+--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
++++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
+@@ -923,12 +923,15 @@ public class ServerPlayer extends Player {
+ 
+         net.kyori.adventure.text.Component deathMessage = event.deathMessage() != null ? event.deathMessage() : net.kyori.adventure.text.Component.empty(); // Paper - Adventure
+ 
+-        if (deathMessage != null && deathMessage != net.kyori.adventure.text.Component.empty() && flag) { // Paper - Adventure // TODO: allow plugins to override?
+-            Component ichatbasecomponent = PaperAdventure.asVanilla(deathMessage); // Paper - Adventure
+-
+-            this.connection.send(new ClientboundPlayerCombatKillPacket(this.getCombatTracker(), ichatbasecomponent), PacketSendListener.exceptionallySend(() -> {
++        // Paper start - death screen message api
++        net.kyori.adventure.text.Component deathScreenMessage = event.deathScreenMessage() != null ? event.deathScreenMessage() : deathMessage;
++        if ((true || deathMessage != null && deathMessage != net.kyori.adventure.text.Component.empty()) && flag) { // Paper - Adventure
```

Comment:
This diff is rather large for how little it needs to do especiall due to the change in control flow that i would personally not suggest.

It might be worth simply evaluating the deathScreenMessage before the branching control flow at the top, similar to how `net.kyori.adventure.text.Component deathMessage` is computed.

To also include `flag` or well the gamerule, it could be computed as follows:
`net.kyori.adventure.text.Component deathScreenMessage = event.deathScreenMessage() != null && flag ? event.deathScreenMessage() : net.kyori.adventure.text.Component.empty(); // Paper - death screen api`

Leaving it empty if either no death message is supposed to be displayed or a plugin set the component to null.
From there on out, we can literally just replace the usages both inside the following if block and in the else block with the variable.

This should get us to a diff such as
```diff
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 289429eb4..b31d359b4 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -922,13 +922,14 @@ public class ServerPlayer extends Player {
         }

         net.kyori.adventure.text.Component deathMessage = event.deathMessage() != null ? event.deathMessage() : net.kyori.adventure.text.Component.empty(); // Paper - Adventure
+        var deathScreenMessage = PaperAdventure.asVanilla(event.deathScreenMessage() != null && flag ? event.deathScreenMessage() : net.kyori.adventure.text.Component.empty()); // Paper - death screen api

         if (deathMessage != null && deathMessage != net.kyori.adventure.text.Component.empty() && flag) { // Paper - Adventure // TODO: allow plugins to override?
             Component ichatbasecomponent = PaperAdventure.asVanilla(deathMessage); // Paper - Adventure

-            this.connection.send(new ClientboundPlayerCombatKillPacket(this.getCombatTracker(), ichatbasecomponent), PacketSendListener.exceptionallySend(() -> {
+            this.connection.send(new ClientboundPlayerCombatKillPacket(this.getCombatTracker(), deathScreenMessage), PacketSendListener.exceptionallySend(() -> {
                 boolean flag1 = true;
-                String s = ichatbasecomponent.getString(256);
+                String s = deathScreenMessage.getString(256);
                 MutableComponent ichatmutablecomponent = Component.translatable("death.attack.message_too_long", Component.literal(s).withStyle(ChatFormatting.YELLOW));
                 MutableComponent ichatmutablecomponent1 = Component.translatable("death.attack.even_more_magic", this.getDisplayName()).withStyle((chatmodifier) -> {
                     return chatmodifier.withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, ichatmutablecomponent));
@@ -948,7 +949,7 @@ public class ServerPlayer extends Player {
                 this.server.getPlayerList().broadcastSystemMessage(ichatbasecomponent, false);
             }
         } else {
-            this.connection.send(new ClientboundPlayerCombatKillPacket(this.getCombatTracker(), CommonComponents.EMPTY));
+            this.connection.send(new ClientboundPlayerCombatKillPacket(this.getCombatTracker(), deathScreenMessage));
         }
```
which achieves the same thing.

Reply:
And I thought my diff was already small enough... :D Thank you, rebased on master and fixed!